Updated documentation (#783)
dynamoosejs/dynamoose: docs/api/Model.md

# Model

The Model object represents one DynamoDB table. It takes in both a name and a schema and has methods to retrieve, and save items in the database.
The Model object represents a table in DynamoDB. It takes in both a name and a schema and has methods to retrieve, and save items in the database.

## new dynamoose.Model(name, schema[, config])

> This method is the basic entry point for creating a model in Dynamoose. When you call this method a new model is created, and it returns a Document initializer that you can use to create instances of the given model.

Reviewer: Would like to hear more about the reasoning for removing `Document` from the language here. I think it's important to have a reference that it's a Document initializer. That way it's clear that functions like `document.save` work correctly. It doesn't just create an initializer, it creates a Document initializer.

Author: I think it was when I started naming document to item. This can be reverted.

Author: Reverted

> The `schema` parameter can either be an object OR a Schema instance, see [Schema](Schema.md) for details. If you pass in an object for the `schema` parameter it will create a Schema instance for you automatically.

Reviewer: No need for extra words here.

```suggestion
The `schema` parameter can either be an object OR a [Schema](Schema.md) instance. If you pass in an object for the `schema` parameter it will create a Schema instance for you automatically.
```

The `schema` parameter can either be an object OR a Schema instance. If you pass in an object for the `schema` parameter it will create a Schema instance for you automatically.
The `schema` parameter can either be an object OR a Schema instance, see [Schema](Schema.md) for details. If you pass in an object for the `schema` parameter it will create a Schema instance for you automatically.

```js

> const dynamoose = require("dynamoose");

Reviewer: Why was this removed? Model is a common starting point for Dynamoose. I think it's important to understand how this is happening.

Author: It is already part of General.md and i didnt see the require anywhere else. So i thought it would consistent to have it nowhere. For me schema would be the entry point.

Reviewer: Adding this back to only the Model and Schema initialization sections.

> const Cat = new dynamoose.Model("Cat", {"name": String}, { create: false });

Reviewer: Any reason for adding this? Right below we go over the `config` parameter. Why introduce it here? And why just `create`?

```suggestion
const Cat = new dynamoose.Model("Cat", {"name": String}, {"create": false});
```

Author: I think it was just to show that a Model accepts options. Can be removed if you think it is not necessary. I know the function definition is in the title, but sometimes one overreads it in the title and then sees the example and thinks it has no config options.

Author: You used `create: false` in a lot of examples so I did that, too for consistency.

> const Cat = new dynamoose.Model("Cat", new dynamoose.Schema({"name": String}), { create: false });

Reviewer: Same question as above.

Reviewer: ```suggestion
const Cat = new dynamoose.Model("Cat", new dynamoose.Schema({"name": String}), {"create": false});
```

| waitForActive.check.timeout | How many milliseconds before Dynamoose should timeout and stop checking if the table is active. | Number | 180000 |
| waitForActive.check.frequency | How many milliseconds Dynamoose should delay between checks to see if the table is active. If this number is set to 0 it will use `setImmediate()` to run the check again. | Number | 1000 |
| update | If Dynamoose should update the capacity of the existing table to match the model throughput. | Boolean | false |

> | expires | The setting to describe the time to live for items created. If you pass in a number it will be used for the `expires.ttl` setting, with default values for everything else. If this is `null`, no time to live will be active on the model. | Number \| Object | null |
> | expires.ttl | The default amount of time the item should stay alive from creation time in milliseconds. | Number | null |
> | expires.attribute | The attribute name for where the item time to live attribute. | String | `ttl` |

Reviewer: ```suggestion
| expires | The setting to describe the time to live for documents created. If you pass in a number it will be used for the `expires.ttl` setting, with default values for everything else. If this is `null`, no time to live will be active on the model. | Number \| Object | null |
| expires.ttl | The default amount of time the document should stay alive from creation time in milliseconds. | Number | null |
| expires.attribute | The attribute name for where the document time to live attribute. | String | `ttl` |
```

}
```

> ## Model.get(primaryKey[, settings][, callback])

Reviewer: ```suggestion
## Model.get(key[, settings][, callback])
```

> You can use Model.get to retrieve a item from DynamoDB. This method uses the `getItem` DynamoDB API call to retrieve the object.

Reviewer: ```suggestion
You can use Model.get to retrieve a document from DynamoDB. This method uses the `getItem` DynamoDB API call to retrieve the object.
```

> `keys` can be a string representing the hashKey/partition key or an object containing the hashKey/partition key AND rangeKey/sort key.

Reviewer: ```suggestion
The `key` parameter can be a string representing the hashKey or an object containing the hashKey and rangeKey.
```

Reviewer: Actually. This should be deleted. It's covered below. It's repetitive.

```suggestion
```

> 

Reviewer: ```suggestion
```

> ## Model.get(hashKey[, settings][, callback])

Reviewer: ```suggestion
## Model.get(key[, settings][, callback])
```

See suggestion below.

});
```

> In the event you have a `rangeKey` for your model, you can pass in an object for the `key` parameter which includes the `hashKey` & `rangeKey`.

Reviewer: Think this was my bad. Want it to be consistent throughout.

```suggestion
In the event you have a rangeKey for your model, you can pass in an object for the `key` parameter which includes the hashKey & rangeKey.
```

});
```

> In the event you have a rangeKey/sort key for your model, you can pass in an object for the `key` parameter which includes the range key/sort key & the hash key/partition key.

Reviewer: ```suggestion
In the event you have a `rangeKey` for your model, you can pass in an object for the `key` parameter which includes the `hashKey` & `rangeKey`.
```

});
```

> In the event you have a rangeKey (sort key) for your model, you can pass in an object for the `hashKey` (partition key) parameter.

Reviewer: I think this should read something like the following.

```suggestion
In the event you have a rangeKey/sort key for your model, you can pass in an object for the `key` parameter which includes the range key/sort key & the hash key/partition key.
```

Then line 127 should be changed to the suggestion above.

Author: > Few suggestions. Still not sure I'm on board about the `Item` vs `Document` thing yet. I think that should maybe be a separate PR and we rename every instance of it.

Yeah lets do that in a separate step!

});
```

> ## Model.batchGet(primaryKeys[, settings][, callback])

Reviewer: ```suggestion
## Model.batchGet(keys[, settings][, callback])
```

> You can use Model.batchGet to retrieve multiple items from DynamoDB. This method uses the `batchGetItem` DynamoDB API call to retrieve the object.

Reviewer: ```suggestion
You can use Model.batchGet to retrieve multiple documents from DynamoDB. This method uses the `batchGetItem` DynamoDB API call to retrieve the object.
```

> 
> `primaryKeys` can be an array of strings representing the hashKey/partition key and/or an array of objects containing the hashKey/partition key AND rangeKey/sort key.

Reviewer: ```suggestion
```

});
```

> In the event you have a range key/sort key for your model, you can pass in an object for the `key` parameter which includes the range key/sort key & the hash key/partition key.

Reviewer: ```suggestion
In the event you have a `rangeKey` for your model, you can pass in an object for the `key` parameter which includes the `hashKey` & `rangeKey`.
```

});
```

> In the event you have a rageKey (sort key) for your model, you can pass in an object for the `hashKey` (partition key) parameter.

Reviewer: This should read the same as above.

```suggestion
In the event you have a range key/sort key for your model, you can pass in an object for the `key` parameter which includes the range key/sort key & the hash key/partition key.
```

## Model.create(document, [settings], [callback])

> This function lets you create a new item for a given model. This function is almost identical to creating a new item and calling `document.save`, with one key difference, this function will default to setting `overwrite` to false.

Reviewer: ```suggestion
This function lets you create a new document for a given model. This function is almost identical to creating a new document and calling `document.save`, with one key difference, this function will default to setting `overwrite` to false.
```

## Model.batchPut(documents, [settings], [callback])

> This saves items to DynamoDB. This method uses the `batchWriteItem` DynamoDB API call to store your objects in the given table associated with the model. This method is overwriting, and will overwrite the data you currently have in place for the existing key for your table.

Reviewer: ```suggestion
This saves documents to DynamoDB. This method uses the `batchWriteItem` DynamoDB API call to store your objects in the given table associated with the model. This method is overwriting, and will overwrite the data you currently have in place for the existing key for your table.
```

});
```

## Model.update(key[, updateObj[, settings]],[ callback])

This function lets you update an existing item in the database. You can either pass in one object combining both the hashKey (partition key) you wish to update along with the update object, or keep them separate by passing in two objects.

> `key` can be a string representing the hashKey or an object containing the hashKey AND rangeKey.

Reviewer: ```suggestion
`key` can be a string representing the hashKey or an object containing the hashKey & rangeKey.
```

The `validate` Schema attribute property will only be run on `$SET` values. This is due to the fact that Dynamoose is unaware of what the existing value is in the database for `$ADD` properties.

## Model.delete(key[, settings][, callback])

You can use Model.delete to delete a item from DynamoDB. This method uses the `deleteItem` DynamoDB API call to delete the object.

> `key` can be a string representing the hashKey or an object containing the hashKey AND rangeKey.

Reviewer: ```suggestion
`key` can be a string representing the hashKey or an object containing the hashKey & rangeKey.
```

> ## Model.delete(hashKey[, settings][, callback])

Reviewer: ```suggestion
## Model.delete(key[, settings][, callback])
```

});
```

> In the event you have a rangeKey for your model, you can pass in an object for the `key` parameter which includes the rangeKey & the hashKey.

Reviewer: ```suggestion
In the event you have a rangeKey for your model, you can pass in an object for the `key` parameter which includes the rangeKey & hashKey.
```

});
```

> In the event you have a rageKey (sort key) for your model, you can pass in an object for the `hashKey` (partition key) parameter.

Reviewer: ```suggestion
In the event you have a range key/sort key for your model, you can pass in an object for the `key` parameter which includes the range key/sort key & the hash key/partition key.
```

## Model.batchDelete(keys[, settings][, callback])

You can use Model.batchDelete to delete items from DynamoDB. This method uses the `batchWriteItem` DynamoDB API call to delete the objects.

> `keys` can be an array of strings representing the hashKey and/or an array of objects containing the hashKey AND rangeKey.

Reviewer: ```suggestion
`keys` can be an array of strings representing the hashKey and/or an array of objects containing the hashKey & rangeKey.
```

});
```

> In the event you have a rageKey (sort key) for your model, you can pass in an object for the `hashKey` (partition key) parameter.

Reviewer: ```suggestion
In the event you have a range key/sort key for your model, you can pass in an object for the `key` parameter which includes the range key/sort key & the hash key/partition key.
```

});
```

> In the event you have a rangeKey for your model, you can pass in an object for the `key` parameter which includes the rangeKey & the hashKey.

Reviewer: ```suggestion
In the event you have a rangeKey for your model, you can pass in an object for the `key` parameter which includes the rangeKey & hashKey.
```

## Model.methods.document.set(name, function)

> This function allows you to add a method to the model items that you can call later. When Dynamoose calls your `function` parameter, `this` will be set to the underlying item. If an existing method exists with the given name, it will be overwritten, except if you are trying to replace an internal method, then this function will fail silently.

Reviewer: ```suggestion
This function allows you to add a method to the model documents that you can call later. When Dynamoose calls your `function` parameter, `this` will be set to the underlying document. If an existing method exists with the given name, it will be overwritten, except if you are trying to replace an internal method, then this function will fail silently.
```

## Model.methods.document.delete(name)

> This allows you to delete an existing method from the item. If no existing method is assigned for that name, the function will do nothing and no error will be thrown.

Reviewer: ```suggestion
This allows you to delete an existing method from the document. If no existing method is assigned for that name, the function will do nothing and no error will be thrown.
```
